compiler/ast.nim

```diff
@@ -1002,12 +1057,152 @@ type
+  n.state.nodeFlag[n.idx] = flags
+
+func intVal*(n: PNode): var BiggestInt {.inline.} =
+  assert n.kind in {nkCharLit..nkUInt64Lit}, "not an integer, id: " & $n.id
+  n.state.nodeInt[n.id]
+proc `intVal=`*(n: PNode, v: BiggestInt) {.inline.} =
+  assert n.kind in {nkCharLit..nkUInt64Lit}, "not an integer, id: " & $n.id
+  n.state.nodeInt[n.id] = v
+
+func floatVal*(n: PNode): BiggestFloat {.inline.} =
+  assert n.kind in {nkFloatLit..nkFloat128Lit}, "not a float, id: " & $n.id
+  n.state.nodeFlt[n.id]
+proc `floatVal=`*(n: PNode, v: BiggestFloat) {.inline.} =
+  assert n.kind in {nkFloatLit..nkFloat128Lit}, "not a float, id: " & $n.id
+  n.state.nodeFlt[n.id] = v
+
+func strVal*(n: PNode): var string {.inline.} =
+  assert n.kind in {nkStrLit..nkTripleStrLit}, "not a string, id: " & $n.id
+  n.state.nodeStr[n.id]
+proc `strVal=`*(n: PNode, v: string) {.inline.} =
+  assert n.kind in {nkStrLit..nkTripleStrLit}, "not a string, id: " & $n.id
+  n.state.nodeStr[n.id] = v
+
+func sym*(n: PNode): PSym {.inline.} =
+  assert n.kind == nkSym, "not a symbol, id: " & $n.id
+  n.state.nodeSym[n.id]
+proc `sym=`*(n: PNode, sym: PSym) {.inline.} =
+  assert n.kind == nkSym, "not a symbol, id: " & $n.id
+  n.state.nodeSym[n.id] = sym
+
+func ident*(n: PNode): PIdent {.inline.} =
+  assert n.kind == nkIdent, "not an ident, id: " & $n.id
+  n.state.nodeIdt[n.id]
+proc `ident=`*(n: PNode, ident: PIdent) {.inline.} =
+  assert n.kind == nkIdent, "not an ident, id: " & $n.id
+  n.state.nodeIdt[n.id] = ident
+
+const haveNoSons = {
+    nkCharLit..nkUInt64Lit,
+    nkFloatLit..nkFloat128Lit,
+    nkStrLit..nkTripleStrLit,
+    nkSym,
+    nkIdent
+  }
+
+template sons*(n: PNode): var TNodeSeq =
+  assert n.kind notin haveNoSons, "not a parent, id: " & $n.id
+  assert n.id.int <= n.state.nodeList.len, "invalid node id: " & $n.id.int
+  # if n.state == nil:
+  #   debugEcho "broken node: ", n.id.int
+  n.state.astData.mgetOrPut(n.id, @[])
+proc `sons=`*(n: PNode, sons: TNodeSeq) {.inline.} =
+  assert n.idx < state.nodeList.len
+  assert n.id.int <= n.state.nodeList.len, "invalid node id: " & $n.id.int
+  n.state.astData[n.id] = sons
+
+proc idToNode*(id: NodeId): PNode {.inline.} =
+  assert id.int <= state.nodeList.len, "invalid node id: " & $id.int
+  case id
+  of nilNodeId: nilPNode
+  else: PNode(id: id, state: state)
+
+proc validateProcLike*(n: PNode) {.inline, deprecated: "only for debugging".} =
+  if not n.state.astData[n.id].len >= 7:
+    echo "invalid proc like node, id: ", n.id
+
+#------------------------------------------------------------------------------
+# DOD AST Draft - End
+#------------------------------------------------------------------------------
 
 type Gconfig = object
   # we put comments in a side channel to avoid increasing `sizeof(TNode)`, which
   # reduces memory usage given that `PNode` is the most allocated type by far.
-  comments: Table[int, string] # nodeId => comment
+  comments: Table[NodeId, string] # nodeId => comment
```

Comment:
Pretty sure you have already though about this, but we can put comments in  the AST state as well, and make all comment accessor functions side-effect free as well

Reply:
I did, just wanted to get things working first, but the VM is being a jerk.